### ISSUE TITLE
Cannot add a user message after another user message (Parameter message

### ISSUE THREAD

**Issue by xuzeyu91** [SciSharp/LLamaSharp#585]
<img width="983" alt="eb7d175bad4438f8595497d76bec183" src="https://github.com/SciSharp/LLamaSharp/assets/26290929/e6ccda12-56ee-4dba-a7e7-8655a0937eb4">

Occasionally appearing "Cannot add a user message after another user message (Parameter message "
Suspected to be caused by concurrent calls

**Comment by AsakusaRinne** [comment]
Hi, thank you for reporting us it. Could you please provide some steps to reproduce it?

**Comment by xuzeyu91** [comment]
https://github.com/xuzeyu91/AntSK/blob/main/src/AntSK/Services/LLamaSharp/LLamaChatService.cs
This is my code, and I use streaming output. If I initiate another request before completing it, this problem will occur

**Comment by xuzeyu91** [comment]
Does this mean it can only be operated with a single thread

**Comment by wrexbe** [comment]
Most things only work with a single thread. If you want to share things between requests, they need to be designed to work that way.

**Comment by AsakusaRinne** [comment]
Seems to be related with #595. Could you please try again on the current master branch?

**Comment by wrexbe** [comment]
Both these methods are not thread safe, it's not going to work reliably if they are going through ChatSession to do the calls.

https://github.com/SciSharp/LLamaSharp/blob/9ee6ae331918d52f7887e3285361428943fb093b/LLama/ChatSession.cs#L234
https://github.com/SciSharp/LLamaSharp/blob/9ee6ae331918d52f7887e3285361428943fb093b/LLama/ChatSession.cs#L390

**Comment by AsakusaRinne** [comment]
Yes, the callings of native apis are thread-safe now but `ChatSession` doesn't. I misunderstood this issue. Do you have any idea about how to make `ChatSession` thread-safe?

**Comment by wrexbe** [comment]
Making it thread safe doesn't really make sense.

The normal scenario is probably going to be 1 user per 1 chat session, and in that case, it's better to just reject the call until the AI is done chatting. You can see something like this happen if you use ChatGPT, and try to send a message while it's still talking.

Calling Chat could detect that it's in the middle of making a response, and throw an InvalidOperationException while it's still busy, so the error message is clearer. The alternatives of feeding messages to Chat Session in a queue, or interrupting the currently generated message will probably need a custom enough solution that they need to use the executor directly anyways.
